File: problems/0010_regular-expression-matching/backtrace/Solution.swift
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        return _isMatch(Array(s), Array(p))
    }
    func _isMatch(_ s: [Character], _ p: [Character]) -> Bool {
        if p.count == 0 {
            return (s.count == 0)
        }
        let p0 = p[0]
        let matchP0 = (s.count > 0 &&
                (p0 == s[0] || p0 == "."))
        if p.count >= 2 && p[1] == "*" {
            return (_isMatch(s, Array(p[2..<p.count])) ||
                    (matchP0 && _isMatch(Array(s[1..<s.count]), p)))
        } else {
            return (matchP0 && _isMatch(Array(s[1..<s.count]), Array(p[1..<p.count])))
        }
    }
}

File: problems/0010_regular-expression-matching/backtrace/Solution.php
<?php
class Solution {
    /**
     * @param String $s
     * @param String $p
     * @return Boolean
     */
    function isMatch($s, $p) {
        $sLen = strlen($s);
        $pLen = strlen($p);
        if ($pLen == 0) return ($sLen == 0);
        $p0 = $p[0];
        $matchP0 = ($sLen > 0 &&
                ($p0 == $s[0] || $p0 == '.'));
        if ($pLen >= 2 && $p[1] == '*')
            return ($this->isMatch($s, substr($p, 2, $pLen - 2)) ||
                    ($matchP0 && $this->isMatch(substr($s, 1, $sLen - 1), $p)));
        else
            return ($matchP0 && $this->isMatch(substr($s, 1, $sLen - 1), substr($p, 1, $pLen - 1)));
    }
}
